File: backend/Testing/Browser/Page/External/MailDevPages.cs
using Microsoft.Playwright;

namespace Testing.Browser.Page.External;

public class MailDevInboxPage : MailInboxPage
{
    public MailDevInboxPage(IPage page, string mailboxId)
    : base(page, "http://localhost:1080/#/", page.Locator("ul.email-list"), mailboxId, page.Locator("ul.email-list li a"))
    {
    }

    protected override MailEmailPage GetEmailPage()
    {
        return new MailDevEmailPage(Page);
    }

    public override async Task<MailInboxPage> Goto()
    {
        await base.Goto();
        await Page.Locator("input.search-input").FillAsync(MailboxId);
        return this;
    }
}

public class MailDevEmailPage : MailEmailPage
{
    public MailDevEmailPage(IPage page) : base(page,
        null, page.FrameLocator(".preview-iframe:visible").Locator("body"))
    {
    }
}

File: frontend/tests/pages/mailDevPages.ts
import type { Page } from '@playwright/test';
import { MailEmailPage, MailInboxPage } from './mailPages';

export class MailDevInboxPage extends MailInboxPage {
  constructor(page: Page, mailboxId: string) {
    super(page,
          page.locator('ul.email-list'),
          'http://localhost:1080/#/',
          mailboxId,
          page.locator('ul.email-list li a'));
  }

  override getEmailPage(): MailEmailPage {
    return new MailDevEmailPage(this.page);
  }

  override async goto(options?: {expectRedirect: boolean}): Promise<this> {
    await super.goto(options);
    await this.page.locator('input.search-input').fill(this.mailboxId);
    return this;
  }
}

export class MailDevEmailPage extends MailEmailPage {
  constructor(page: Page) {
    super(page, page.frameLocator('.preview-iframe:visible').locator('body'), undefined);
  }
}
